src/logic/map_objects/tribes/ship.cc

```diff
@@ -52,49 +52,62 @@ namespace Widelands {
 
 namespace {
 
-/// Returns true if 'coord' is not occupied or owned by 'player_number' and
-/// nothing stands there.
-bool can_support_port(const PlayerNumber player_number, const FCoords& coord) {
-	const PlayerNumber owner = coord.field->get_owned_by();
-	if (owner != neutral() && owner != player_number) {
+/// Returns true if 'coord' is not blocked by immovables
+/// Trees are allowed, because we don't want spreading forests to block portspaces from expeditions
+bool can_support_port(const FCoords& coord) {
+	BaseImmovable* baim = coord.field->get_immovable();
+	Immovable* imo = dynamic_cast<Immovable*>(baim);
+	// we have a player immovable
+	if (imo == nullptr && baim != nullptr) {
 		return false;
 	}
-	BaseImmovable* baim = coord.field->get_immovable();
-	return (baim == nullptr || baim->descr().type() < MapObjectType::FLAG);
+	return (baim == nullptr || baim->get_size() < BaseImmovable::SMALL ||
+	        imo->descr().has_terrain_affinity());
 }
 
 /// Returns true if a ship owned by 'player_number' can land and erect a port at 'coord'.
 bool can_build_port_here(const PlayerNumber player_number, const Map& map, const FCoords& coord) {
-	if (!can_support_port(player_number, coord)) {
-		return false;
-	}
-
+	// First check ownership of the port space
 	// All fields of the port + their neighboring fields (for the border) must
-	// be conquerable without military influence.
-	Widelands::FCoords c[4];  //  Big buildings occupy 4 locations.
+	// be conquerable without military influence. Check radius 2 around
+	// the main spot to cover radius 1 around each of the 4 fields and the flag
+	MapRegion<Area<FCoords>> area(map, Area<FCoords>(coord, 2));
+	do {
+		const PlayerNumber owner = area.location().field->get_owned_by();
+		if (owner != neutral() && owner != player_number) {
+			return false;
+		}
+	} while (area.advance(map));
+
+	// then check whether we can build a port like if reached from land
+	if ((coord.field->nodecaps() & BUILDCAPS_SIZEMASK) == BUILDCAPS_BIG &&
+	    !map.find_portdock(coord, false).empty()) {
+		return true;
+	}
+	// now check if we are allowed to build one although some conditions not met
+	// All fields of the port and some neighbouring fields must be free of
+	// blocking immovables.
+	Widelands::FCoords c[14];
 	c[0] = coord;
 	map.get_ln(coord, &c[1]);
 	map.get_tln(coord, &c[2]);
 	map.get_trn(coord, &c[3]);
-	for (const Widelands::FCoords& fc : c) {
-		MapRegion<Area<FCoords>> area(map, Area<FCoords>(fc, 1));
-		do {
-			if (!can_support_port(player_number, area.location())) {
-				return false;
-			}
-		} while (area.advance(map));
-	}
+	map.get_rn(coord, &c[4]);
+	map.get_brn(coord, &c[5]);
+	map.get_bln(coord, &c[6]);
+	map.get_bln(c[0], &c[7]);
+	map.get_ln(c[0], &c[8]);
+	map.get_tln(c[0], &c[9]);
+	map.get_tln(c[1], &c[10]);
+	map.get_trn(c[1], &c[11]);
+	map.get_trn(c[2], &c[12]);
+	map.get_rn(c[2], &c[13]);
```

Comment:
I found a mistake in https://github.com/widelands/widelands/pull/5647/commits/e856f8e7a72eb4f383813e3ec11f8c0677df8ca0 doing my usual tests. Base coordinate indexes were not increased when main `coord` was reinserted in `c[]`.

```suggestion
	map.get_bln(c[1], &c[7]);
	map.get_ln(c[1], &c[8]);
	map.get_tln(c[1], &c[9]);
	map.get_tln(c[2], &c[10]);
	map.get_trn(c[2], &c[11]);
	map.get_trn(c[3], &c[12]);
	map.get_rn(c[3], &c[13]);
```

Reply:
So full test coverage would need trying with a rock or other size 1 immovable at each of these 14 positions one by one, but do we need or want that?

Reply:
I don't think so, cause to test all permutations we might need approx more then 50 tests. as we might need to test with size 1 immos or size 3 (like rocks as they are big actually and block more space).